xls/ir/nodes_source.tmpl

```diff
@@ -107,6 +108,36 @@ SliceData Concat::GetOperandSliceData(int64 operandno) const {
   return SliceData{start, this->operand(operandno)->BitCountOrDie()};
 }
 
+int64 ArrayIndex::GetAddressableElementCount() const {
+  int64 element_product = 1;
+  Type* type = array()->GetType();
+  for(const Node* index_node : indices()) {
+    ArrayType* array_type = type->AsArrayOrDie();
+    int64 bits_to_address_all_elements =
+      Bits::MinBitCountUnsigned(array_type->size());
+    int64 index_bit_count = index_node->GetType()->GetFlatBitCount();
+    element_product *= index_bit_count < bits_to_address_all_elements ?
```

Comment:
I think the fact that the addressableelementcount depends upon the width of the index types is surprising. For example, replacing a constant index with a equivalent but narrower constant will change this value (and a constant indexes a single value actually which isn't reflected here). Anyway, I think the marginal value added for more precise modeling of array index and array update is not outweighed by the surprising behavior of this function. I think just using the number of subelements in the indexed array (e.g, array flat bit count  / output flat bit count) would be better.

I think more precise modelling is better handled through range and bit value analysis rather than a method on a Node class. Such more complex analysis might tell you the actual ranges of the indices and whether certain bits are zero/one.

Reply:
Done.